### ISSUE TITLE
Allow `keep_unary` to be passed to simplification funcs

### ISSUE THREAD

**Issue by hyanwong** [MesserLab/SLiM#139]
I can imagine that it would be very handy to be able to output in  tree sequence all the intermediate nodes produced by SLiM - i.e. not just the coalescence points, but all the "unary nodes" too. This would be somewhat intermediate between "saving everybody" and the normal output. Individuals would still not be saved, of course, unless they happen to be ["retained"](#40) - this is just about saving the node info (genomes) themselves.

The way I imagine this working would be to add an optional `keep_unary` arg to `treeSeqSimplify()`, which could be T, F, or NULL (default). There would also be a `keep_unary` option to `initializeTreeSeq()` that would default to `F` and would set the default `keep_unary` flag during automatic simplification. This value would be the default used when calling `treeSeqSimplify()` with no params.

What do you think @petrelharp ? I guess you're the one doing "remember everyone" simulations.

**Comment by bhaller** [comment]
As I mentioned in the review on #25, let's try to envision the ideal final design for this stuff.  Is it a `keepUnary=T` option on `initializeTreeSeq()` and a `permanent=F` option on `treeSeqRememberIndividuals()`?  In other words, are these two features completely orthogonal?  Or is there some interaction between them, and some better API design?  Because `keepUnary=T` strikes me as a kind of "remembering" or "retaining" too, but we're proposing handling it in a very different way.  It could be handled instead as another type of flag passed to `treeSeqRememberIndividuals()`, right?  A flag saying "mark these individuals as being kept even if their nodes are unary"?  I don't see a clear problem with what has been proposed, but I want to make sure we're going to arrive at a final design that we won't regret later.

**Comment by hyanwong** [comment]
I see them as completely orthogonal, which is why I opened up a new issue. I guess you could think of this as a sort of "remembering", but of *nodes* not individuals. The distinction is very clear in tree-sequence world, and I think that the "keep_unary" function should automatically play nicely (and logically) with the `permanent=F` option in #25 . But let's see what @petrelharp thinks too?

**Comment by petrelharp** [comment]
Not even so much orthogonal, but copmplementary: if you do `keep_unary=T` and you Retain all the individuals, then you'll get the individuals corresponding to retained unary nodes as well. (However, I'd imagine that would be nearly the same as just Remembering everyone.)

**Comment by hyanwong** [comment]
> (However, I'd imagine that would be nearly the same as just Remembering everyone.)

I think it would be for short simulations. For long running ones, I suspect that there will be lots of ancestors in the distant past that are genealogical ancestors but not genetic ancestors anywhere (unless e.g. there is loads of gene conversion)

**Comment by hyanwong** [comment]
Oh, one extra wrinkle I just thought of. There's not much point to allowing `treeSeqSimplify(keep_unary=T)` if by default the next time that simplification is done automatically, it removes all the unary nodes anyway, because we started the simulation with `initializeTreeSeq(keep_unary=F)`. But unless anyone has a better suggestion, I guess there's no *harm* in allowing it, as long as it's documented that subsequent auto-simplification could remove the unary nodes.

**Comment by bhaller** [comment]
> Oh, one extra wrinkle I just thought of. There's not much point to allowing `treeSeqSimplify(keep_unary=T)` if by default the next time that simplification is done automatically, it removes all the unary nodes anyway, because we started the simulation with `initializeTreeSeq(keep_unary=F)`. But unless anyone has a better suggestion, I guess there's no _harm_ in allowing it, as long as it's documented that subsequent auto-simplification could remove the unary nodes.

Why even have it as a parameter to `treeSeqSimplify()`?  Why not just use the flag that was given to `initializeTreeSeq()` for all simplifications?

**Comment by bhaller** [comment]
> > (However, I'd imagine that would be nearly the same as just Remembering everyone.)
> 
> I think it would be for short simulations. For long running ones, I suspect that there will be lots of ancestors in the distant past that are genealogical ancestors but not genetic ancestors anywhere (unless e.g. there is loads of gene conversion)

OK, probably I'm just confused, but.  Remembering all individuals would mean that whole evolutionary branches that eventually went extinct would still be preserved in the tree sequence, right?  *Every* individual would be remembered.  Whereas retaining every individual and setting `keepUnary=T` would keep every individual that is genetically ancestral to the extant generation, but it would allow the evolutionary branches that went extinct to be pruned away by simplify, right?  So isn't there a huge difference between the two, given that most evolutionary branches eventually go extinct?  Or what am I misunderstanding?

**Comment by petrelharp** [comment]
If we `keepUnary`, then we'll prune out only individuals who leave behind no genome to the present day. If you're doing a long chromosome, that's nearly the same as individuals who have no pedigree ancestors in the present day. The number of pedigree descndants of an individual in a WF population acts, for a while anyhow, like a branching process with a Poisson(2) offspring distribution, whose probability of extinction is about 20%. So, I'd expect if you're doing a simulation with a long genome (like, 1Morgan long), that keepUnary would end up with about 80% as many individuals as compared to if you Remembered everyone. If you're doing a short genome though, then I agree, the number would be a lot smaller.

**Comment by bhaller** [comment]
> If we `keepUnary`, then we'll prune out only individuals who leave behind no genome to the present day. If you're doing a long chromosome, that's nearly the same as individuals who have no pedigree ancestors in the present day.

Yes – but it's much, much smaller than the total number of individuals that have ever lived, right?  Which is what just remembering every individual in every generation would do, right?  I mean, this is the main point of simplification, isn't it?  That over time, most evolutionary branches go completely extinct: most individuals leave no pedigree descendants over the long run.  But remembering every individual in every generation keeps all of those extinct branches, and effectively turns off simplification, doesn't it?  Or am I misremembering how "remembering" works?

**Comment by petrelharp** [comment]
> Yes – but it's much, much smaller than the total number of individuals that have ever lived, right? 

Well, this is an empirical question, but my calculation above suggests that the number of individuals with no pedigree descendants is about 80% of the total number of individuals ever.

I think you're getting mixed up by the fact that the tree sequence only has to keep track of nodes with *coalescences* in extant material, not *all* nodes that are ancestors. So the condition for appearing in a tree sequence is "I'm the MRCA of at least *two* samples at some locus", which is a lot stricter than "I'm an ancestor of at least one sample at some locus".

**Comment by petrelharp** [comment]
Just to be clear, I'm fine with this proposed change, but I suspect that in practice I'm going to keep on remembering everyone.

**Comment by hyanwong** [comment]
> Well, this is an empirical question, but my calculation above suggests that the number of individuals with no pedigree descendants is about 80% of the total number of individuals ever.

I think it's a relatively standard result that the pedigree ancestors encompass about 80% of an entire panmictic population (I think Richard Dawkins and I were the first to write about this in a popular science book, based on Chang's calculations, around the year 2000!). But that's not the same and the percentage of *genetic* (not pedigree) ancestors, which are the ones we would keep if we did `keep_unary`. Are pedigree and genetic ancestors really the same order of magnitude? I can see that's the case as the chromosomes get very long, but I'm not sure it is in e.g. humans. Graham Coop has a nice blog post about this, ISTR, and I asked him about it (?on twitter), assuming that far enough into the past, the genetic lineages get cut up into such fine pieces that they basically pass through all the pedigree ancestors. But he said that doesn't happen - something to do with the asymptotic effect of recombination.

I guess if we implement this feature, it'll be easy to find out!

**Comment by bhaller** [comment]
> > Yes – but it's much, much smaller than the total number of individuals that have ever lived, right?
> 
> Well, this is an empirical question, but my calculation above suggests that the number of individuals with no pedigree descendants is about 80% of the total number of individuals ever.
> 
> I think you're getting mixed up by the fact that the tree sequence only has to keep track of nodes with _coalescences_ in extant material, not _all_ nodes that are ancestors. So the condition for appearing in a tree sequence is "I'm the MRCA of at least _two_ samples at some locus", which is a lot stricter than "I'm an ancestor of at least one sample at some locus".

No, that's not the point of confusion for me, I get that.  The point of confusion for me is that I can't reconcile what you're saying with, e.g., this claim: "Of all species that have existed on Earth, 99.9 percent are now extinct." (https://www.pbs.org/wgbh/evolution/extinction/massext/statement_03.html).  In the long run, as I understand it, the vast, vast majority of evolutionary branches go completely extinct; no individual in those branches has any descendant whatsoever.  If you took a complete tree sequence of the entire history of life on Earth and simplified it, pruning away those extinct branches would get rid of 99.9% of the individuals.  (Well, a few of the extinct species would be ancestral to present-day species and so some of their individuals would be retained; but most would be pruned.)  You would then get *additional* savings by pruning away all the nodes that don't involve coalescences, etc., as you say; but let's ignore that for now, since that's just gravy on top of the 99.9% pruning.  But if you had done a "remember" on every individual in every generation, simplify() wouldn't get rid of any of the individuals at all, as I understand it.

So how does your claim that 80% of the individuals would be kept square with this perspective that suggests to me that 99.9% of the individuals would be pruned away?  What am I missing?  (Apologies for the digression.  :->)

**Comment by petrelharp** [comment]
> I think it's a relatively standard result that the pedigree ancestors encompass about 80% of an entire panmictic population (I think Richard Dawkins and I were the first to write about this in a popular science book, based on Chang's calculations, around the year 2000!).

Hah, that's right. =)

>  "Of all species that have existed on Earth, 99.9 percent are now extinct." 

Sure, but that's a single tree. The same thing is true for the tree at any one particular locus; but here we're taking the union across all the trees.

**Comment by petrelharp** [comment]
> Are pedigree and genetic ancestors really the same order of magnitude as @petrelharp says?

I'm a bit turned around now, but consider: in a pop of size N, consider who a single genome has inherited from back N generations. On a 1M chromosome, the genome will be broken up into about N chunks; each will be like a random draw from the population; if you throw N balls at random in N boxes, the proportion of boxes with no balls in will be exp(-1) = 0.367, suggesting that roughly 36% of the individuals won't be genetic ancestors of this individual; and since N generations ago there's only a few lineages, the number of genetic ancestors of everyone will be similar. I guess that random variation in short-term fitness will increase the number that won't be ancestors... so, here I'm getting an answer of roughly 50% non-genetic ancestors.

This would be a fun thing to work out for reals (or maybe the answer is in Barton & Etheridge).

**Comment by bhaller** [comment]
> > "Of all species that have existed on Earth, 99.9 percent are now extinct."
> 
> Sure, but that's a single tree. The same thing is true for the tree at any one particular locus; but here we're taking the union across all the trees.

No, I'm talking about the union across all trees.  Let's take a concrete example.  There was a branch of mammals called multituberculates.  We know of more than 200 species, from the fossil record.  The branch is entirely extinct; no living species descends from any multituberculate species.  (https://ucmp.berkeley.edu/mammal/multis/multis.html)

This means that this branch of the evolutionary tree has no pedigree descendants, for *any* position along the chromosome.  We're not talking about a single gene tree; we're talking about the entire tree sequence.  A simplify of the tree-sequence-of-life would prune away that entire branch completely (modulo some individuals near the root of the tree that hybridized out to adjacent branches before reproductive isolation from other branches was 100% complete; but that'd be a negligible fraction of the total number of individuals in the entire multituberculate branch).  But if one had been remembering every individual in every generation, the entire branch would be kept (*everything* would be kept).

This is not particularly unusual, it seems to me.  The claim that "Of all species that have existed on Earth, 99.9 percent are now extinct" seems to me to imply that there are quite a lot of branches like this on the tree of life – branches that left no living descendent at the present day, and would be pruned away completely by a simplify() of the tree-sequence-of-life.  And the savings from that seems like it would have to be far more than the 20% you're claiming.  I happily concede that I don't know what I'm talking about and am probably wrong – but *why* am I wrong?  I don't feel like we've gotten to the bottom of that yet.  :->

**Comment by petrelharp** [comment]
Hm, does this help: Relationships between individuals of distinct mammal species looks more or less like a single tree.  Relationships *within* one of those branches, ie, between individuals of the same mammal species, are not described by a single tree: there's very different trees as you move along the genome.  If you zoomed in on the nodes of the tree of mammal species you'd see that in fact there is not a single tree "due to incomplete lineage sorting", but at least for well-separated taxa, all the trees look pretty similar. . ..?

**Comment by hyanwong** [comment]
The 80% figure is from a completely panmictic population. One you get subdivision (in the extreme case, many separate species) then it goes down drastically, potentially to the levels that @bhaller says.

**Comment by bhaller** [comment]
Aha, I see.  Yes, I can see that 80% could be reasonable given an assumption of complete panmixia.  But then that suggests that for many SLiM simulations that strongly depart from panmixia, the number might be much less than 80%, and the savings from retaining rather than remembering in each generation might be correspondingly larger, right?

**Comment by hyanwong** [comment]
Back to to original topic, I think what @mmosmond would really like is the ability to save individuals (and their nodes) at a certain timepoint (like a "census") in the simulation. These nodes should be saved even if they are unary. But other unary nodes need not be (in fact, probably shouldn't be) saved. I think this is a tskit thing: we should be able to flag up a node for retention-even-if-unary during normal simplification, but allow it to be removed if it is no longer in the ancestry. I see this as an extra flag on a node, like `NODE_IS_CENSUS`. But perhaps there's another way to do this. I wonder if this actually does change the calling convention for `treeSeqRememberIndividuals()`, since we might want to "remember permanently", "retain if coalescent" or "retain if ancestral" (the last of which would set the `NODE_IS_CENSUS` flag on the 2 nodes). That would render the "keep unary" suggestion here less relevant for this use case, although it might be handy for other people.

Since this isn't a feature of tskit at the moment, though, it's a bit of a long-term aim. And There might well be a better way to do this.

**Comment by petrelharp** [comment]
> Back to to original topic, I think what @mmosmond would really like is the ability to save individuals (and their nodes) at a certain timepoint (like a "census") in the simulation. 

If it's only at certain times, I don't see any reason to go to the bother of keeping them only if they're still in the ancestry - why not just Remember everyone alive at the time.

**Comment by hyanwong** [comment]
> If it's only at certain times, I don't see any reason to go to the bother of keeping them only if they're still in the ancestry - why not just Remember everyone alive at the time.

I take your point, and indeed that's what he's doing at the moment. I guess it's a bit wasteful if, as we say above, (a) there is a lot of population structure or very few trees and (b) there are a lot of census times that are a long time ago in the simulation, e.g. prior to the grand MRCA. But yes, taking a census at a specific time seems to be relatively well dealt with by remembering with `permanent=T`, and for most normal cases, what I'm suggesting would only provide a very small optimisation.

**Comment by hyanwong** [comment]
Or, I guess, if you have a very large population but only a small handful of sample lineages.

**Comment by hyanwong** [comment]
I've just opened up a tskit issue: https://github.com/tskit-dev/tskit/issues/1113 - the approach we take in SLiM might depend on this? I think `treeSeqRememberIndividuals(permanent=F)` (which we have already done, but not released) is still useful though? Thoughts @petrelharp ?

**Comment by bhaller** [comment]
The approach proposed by Jerome in tskit-dev/tskit#1113 makes sense to me.  The additional win provided by being able to specify particular nodes as being "census" nodes and treating them specially seems rather marginal, compared to just remembering all individuals at a specific time point.  The additional algorithmic complexity might not be worth the payoff, and might also slow down simplification for the standard usage case.  But that's not my bailiwick; if the tskit folks decide that it makes sense to add a "census" node flag, it's easy enough to add the necessary API to SLiM for it.  (But again, I'm a bit worried that we'll end up with a bad API if we keep adding these things in one by one in a design-by-accretion approach.)  It sounds like for now the plan is to provide an overall "keep all ancestors" flag for simplify; that can just be added as a flag to initializeTreeSeq(), I imagine, to provide a policy for all simplifications done during a simulation.  If there's a value-add to providing it as a flag on treeSeqSimplify() as well, we can do that, but the value-add is not clear to me at present – would one ever want to use different policies for different simplifications within a single simulation run?

And: is this something that is going to roll in 3.5.1 – i.e., something that is going to happen more or less immediately?  Since it requires changes on the tskit side, then perhaps not?  Somebody reported the `source()` bug on slim-discuss, so clearly it is biting people; so I'm inclined to roll 3.5.1 ASAP, but if this were going to happen immediately then I'd be happy to roll it in.

**Comment by petrelharp** [comment]
Note that it would be very easy to measure the reductions in file size you'd get from this in practice: for instance, by (a) remembering everyone at the times of interest; (b) un-marking the older nodes as samples; (c) doing simplify(keep_unary=True).

**Comment by bhaller** [comment]
Am I correct in thinking that the decision was that a new flag for simplify() will be added on the tskit side, and once that exists it will be exposed as a new flag in initializeTreeSeq()?  And am I right to think that the tskit work will take a bit of time, and so this feature will not be rolling in SLiM 3.5.1?  I'm hoping to get that out the door ASAP now.

**Comment by petrelharp** [comment]
I think that whichever way we go, this feature is not blocking for anyone, so no need to wait on it. (correct me if I'm wrong, @hyanwong)

**Comment by hyanwong** [comment]
Yes, you are right @petrelharp  - this is not blocking, and it will not remove the utility of having the `permanent=T` flag in the RememberIndividuals function. So I think 3.5.1 can be released as soon as you like @bhaller .

**Comment by hyanwong** [comment]
Following on from https://github.com/tskit-dev/tskit/issues/1113, I wonder if, actually, we simply want `keep_unary_in_individuals=True` to be the default simplification method in SLiM? After all, the only non-alive individuals we will encounter are either Remembered (in which case this makes no difference), or Retained (in which case we probably do want to keep them, even if they end up only having a unary node.

A slight wrinkle: if we do aim to make it a default, this flag doesn't actually exist in simplify yet. So there will be a short period of time in the SLiM releases where Retaining an individual does something different. Does that mean we want to e.g. undocument the `permanent=T` option until tskit has the `simplify(keep_unary_in_individuals=True)` ability?

**Comment by bhaller** [comment]
> Following on from [tskit-dev/tskit#1113](https://github.com/tskit-dev/tskit/issues/1113), I wonder if, actually, we simply want `keep_unary_in_individuals=True` to be the default simplification method in SLiM?

This strikes me as making sense, and I can't see any obvious reason why someone would want the opposite behavior.  If they did want that, they could get it by doing another simplify on the python side, right?  Just strip away the unwanted nodes/individuals by simplifying again with the desired setting, since what they want to keep is a strict subset of what SLiM has kept for them?  What do you think, @petrelharp?

> A slight wrinkle: if we do aim to make `keep_unary_in_individuals` the default, this flag doesn't actually exist in simplify yet. So there will be a short period of time in the SLiM release schedule where Retaining an individual does something different. Does that mean we want to e.g. undocument the `permanent=T` option until tskit has the `simplify(keep_unary_in_individuals=True)` ability? Or shall we just update the docs later to say something like
> 
> > _changes since 3.4.1_: the `keep_unary_in_individuals` switch is now used in simplification, so that Retained individuals with genetic ancestry are kept even if their genomes do not represent a coalescence point in the simulation

I think what this means is that it'd be great to get this added to tskit ASAP, and roll the final version of the behavior into SLiM 3.5.1 so there's no discontinuity.  Am I right in thinking it's quick to implement on the tskit side?  And perhaps we could then migrate tskit's code into SLiM to enable the final version of the feature, even before it has been officially announced in a tskit release?  (Or is that a bad idea for reasons of QA/testing/versioning, etc.?)  If we would want to wait for the next tskit release, when is that likely to happen?  I did want to roll 3.5.1 ASAP, because of the source() bug, but I can talk myself into that not being top priority.  I would like to do this feature right; and pulling the feature for 3.5.1 would be a hassle at this point.  Thoughts?

**Comment by petrelharp** [comment]
> > I wonder if, actually, we simply want keep_unary_in_individuals=True to be the default simplification method in SLiM?
> This strikes me as making sense, and I can't see any obvious reason why someone would want the opposite behavior. 

Right now, if someone wants to know individual-level information about all the nodes in the tree sequence, they can just `retain` everyone; then the tree sequence will be somewhat bigger - like, 20%? - because each node (or two) will have an associated individual. This is really excellent. Retain is really great.  However, `keep_unary` is pretty different. Keeping track of all the unary nodes along the way is on the order of remembering everyone, ever, and negates the efficiencies of the tree sequence. Sometimes that's what we have to do, because we really do want to know where all the lineages were at all the times, but I think it definitely needs to be optional.

I'm less sure what I think the default should be? I agree, if I was retaining-but-not-remembering everone, I'd probably want keep_unary also... but since it has such a gigantic effect on file size and runtime, I guess I think that `keep_unary` needs to be False by default, rather than assuming that people who are retaining individuals know what they're getting into.

**Comment by bhaller** [comment]
> > > I wonder if, actually, we simply want keep_unary_in_individuals=True to be the default simplification method in SLiM?
> > > This strikes me as making sense, and I can't see any obvious reason why someone would want the opposite behavior.
> 
> Right now, if someone wants to know individual-level information about all the nodes in the tree sequence, they can just `retain` everyone; then the tree sequence will be somewhat bigger - like, 20%? - because each node (or two) will have an associated individual. This is really excellent. Retain is really great. However, `keep_unary` is pretty different. Keeping track of all the unary nodes along the way is on the order of remembering everyone, ever, and negates the efficiencies of the tree sequence. Sometimes that's what we have to do, because we really do want to know where all the lineages were at all the times, but I think it definitely needs to be optional.
> 
> I'm less sure what I think the default should be? I agree, if I was retaining-but-not-remembering everone, I'd probably want keep_unary also... but since it has such a gigantic effect on file size and runtime, I guess I think that `keep_unary` needs to be False by default, rather than assuming that people who are retaining individuals know what they're getting into.

I could go either way.  People using these flags are likely to be relatively sophisticated/knowledgeable users, and the documentation should explain the various behaviors very explicitly; so either way we choose, probably the people who care about this will be fine.  But I suppose there's a "least harm" argument to be made.  If we choose the default for `keep_unary_in_individuals` to be F, then people who really want it to be true will immediately see that the nodes/individuals that they want to be kept aren't being kept, and they will read the doc or ask questions, and they'll figure it out.  If we choose the default for `keep_unary_in_individuals` to be T, then people who really want it to be F might just end up with slow runtimes and very large .trees files, and never really understand why, or even realize that they're doing something wrong.  This seems to suggest that a default of F would be better.

That said, I confess that I've gotten a bit confused regarding the various options we've been discussing; I think there have been at least three different options on the table (?), and although I gather the tskit thread has settled on one as being the best, I personally would find it very helpful to have some kind of synopsis, perhaps with diagrams, that summarizes what the options are and how they differ.  :-O  Such diagrams would be super-useful for the doc anyway; I'd love to have a big figure in the SLiM manual with different tree sequences shown vertically, side by side in a 4x1 or 5x1 format, depicting which nodes and which Individual objects get preserved under different modes of operation.  It's all getting rather complicated.

**Comment by hyanwong** [comment]
> But I suppose there's a "least harm" argument to be made ... This seems to suggest that a default of F would be better.

That's a great point @bhaller . Incidentally, if we go with this, then we don't need to wait on any tskit stuff to release 3.5.1. Then later, we can release 3.5.2 with the "keep_unaryXXX" option structured to match whatever is decided in the tskit Python API, in discussion at https://github.com/tskit-dev/tskit/issues/1120. This seems a good way forward to me.

**Comment by petrelharp** [comment]
So, what's the proposal now? To add the abililty to pass the `KEEP_UNARY_IN_INDIVIDUALS` flag to simplify? Or do we *also* want the ability to pass in `KEEP_UNARY` itself? Seems to me the first is probably sufficient, so maybe we should stick to that.

And, let's see: `KEEP_UNARY_IN_INDIVIDUALS` only affects what happens to Retained nodes;  the distinction between retain-plus-keep-unary-in-individuals and remembered is that bits of their genomes not ancestral to any samples won't be kept in the first case; and the distinction between retained and retained-plus-keep-unary-in-individuals is that in the second case the bits of their genome will be kept even if they aren't a MRCA of two samples.

So, how about we call the `KEEP_UNARY_IN_INDIVIDUALS` option, um let's see, `keepAllAncestralRetained=F`? And the documentation could say:
"
The keepAllAncestralRetained option changes how much of the genomes of "retained" indviduals (i.e., those remembered with treeSeqRememberIndividuals( , permanent=F)) persist in the tree sequence through simplification. If this option is F (the default), only those portions of retained individuals on which they are the most recent common ancestor of at least two samples is retained (since this is all that is necessary to reconstruct the trees).  (Recall that the "samples" are the final generation and any permanently remembered individuals.)  If this is set to T, then any portion of their genome that is ancestral to the samples will persist in the tree sequence. If a large number of individuals are retained, this will dramatically increase the size of the tree sequence.  Note that for this to be effective, the corresponding option in treeSeqSimplify must be used as well, and that this corresponds to the `keep_unary_in_individuals` flag in the python `simplify` method.
"

**Comment by bhaller** [comment]
> So, what's the proposal now? To add the abililty to pass the `KEEP_UNARY_IN_INDIVIDUALS` flag to simplify? Or do we _also_ want the ability to pass in `KEEP_UNARY` itself? Seems to me the first is probably sufficient, so maybe we should stick to that.

I have lost track of the different switches and options, and the decisions made on them, as I think I mentioned a while ago.  :-<  Were two options added to simplify() in tskit, KEEP_UNARY_IN_INDIVIDUALS and KEEP_UNARY?  I thought Jerome shot down the options other than KEEP_UNARY_IN_INDIVIDUALS; I take it I'm mistaken?

> And, let's see: `KEEP_UNARY_IN_INDIVIDUALS` only affects what happens to Retained nodes; the distinction between retain-plus-keep-unary-in-individuals and remembered is that bits of their genomes not ancestral to any samples won't be kept in the first case; and the distinction between retained and retained-plus-keep-unary-in-individuals is that in the second case the bits of their genome will be kept even if they aren't a MRCA of two samples.
> 
> So, how about we call the `KEEP_UNARY_IN_INDIVIDUALS` option, um let's see, `keepAllAncestralRetained=F`? And the documentation could say:
> "
> The keepAllAncestralRetained option changes how much of the genomes of "retained" indviduals (i.e., those remembered with treeSeqRememberIndividuals( , permanent=F)) persist in the tree sequence through simplification. If this option is F (the default), only those portions of retained individuals on which they are the most recent common ancestor of at least two samples is retained (since this is all that is necessary to reconstruct the trees). (Recall that the "samples" are the final generation and any permanently remembered individuals.) If this is set to T, then any portion of their genome that is ancestral to the samples will persist in the tree sequence. If a large number of individuals are retained, this will dramatically increase the size of the tree sequence. Note that for this to be effective, the corresponding option in treeSeqSimplify must be used as well, and that this corresponds to the `keep_unary_in_individuals` flag in the python `simplify` method.
> "

Seems good to me, perhaps apart from the last sentence.  I thought we decided that if this switch is flipped to T, then all simplify() calls done by SLiM would use the flag automatically...?  What's the purpose of having the switch in two different places in the API?  What would it mean to specify F in one call and T in the other, or T in one call and F in the other?

Do you guys intend to produce diagrams illustrating how the tree sequence is affected by these different settings (regular vs. retain vs. remember, with/without KEEP_UNARY_IN_INDIVIDUALS), for the tskit or pyslim docs?  If so, I think it would be tremendously useful for such diagrams to appear in the SLiM manual as well, probably right under the text you wrote above.  :->

**Comment by petrelharp** [comment]
>  Were two options added to simplify() in tskit, KEEP_UNARY_IN_INDIVIDUALS and KEEP_UNARY? I thought Jerome shot down the options other than KEEP_UNARY_IN_INDIVIDUALS; I take it I'm mistaken?

The [`KEEP_UNARY` option has been around for some time](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.simplify).

>  I thought we decided that if this switch is flipped to T, then all simplify() calls done by SLiM would use the flag automatically...? What's the purpose of having the switch in two different places in the API? What would it mean to specify F in one call and T in the other, or T in one call and F in the other?

Oh, yes, that's a much better idea.

> Do you guys intend to produce diagrams illustrating how the tree sequence is affected by these different settings (regular vs. retain vs. remember, with/without KEEP_UNARY_IN_INDIVIDUALS), for the tskit or pyslim docs? 

Yep - Yan's got drafts of these already, even. Happy for them to go in the SLiM manual also.

**Comment by bhaller** [comment]
> > Were two options added to simplify() in tskit, KEEP_UNARY_IN_INDIVIDUALS and KEEP_UNARY? I thought Jerome shot down the options other than KEEP_UNARY_IN_INDIVIDUALS; I take it I'm mistaken?
> 
> The [`KEEP_UNARY` option has been around for some time](https://tskit.readthedocs.io/en/latest/python-api.html#tskit.TreeSequence.simplify).

Ah, I didn't know this existed.  :->

> > I thought we decided that if this switch is flipped to T, then all simplify() calls done by SLiM would use the flag automatically...? What's the purpose of having the switch in two different places in the API? What would it mean to specify F in one call and T in the other, or T in one call and F in the other?
> 
> Oh, yes, that's a much better idea.

Ah, good.

> > Do you guys intend to produce diagrams illustrating how the tree sequence is affected by these different settings (regular vs. retain vs. remember, with/without KEEP_UNARY_IN_INDIVIDUALS), for the tskit or pyslim docs?
> 
> Yep - Yan's got drafts of these already, even. Happy for them to go in the SLiM manual also.

Yay!

OK, so.  I'm good with `keepAllAncestralRetained=F` if you guys like that.  Shall I do a commit that adds this parameter to initializeTreeSeq(), and puts the value into a flag in SLiMSim for your use, and adds the doc above to the online doc in SLiMgui?  I'm fine with just committing that and having it be a loose end for a couple days, until it gets hooked up to something real by a PR from you guys.  Nobody will notice that it's even there, I imagine, and it won't break anything.  I could do that now-ish if that sounds like a good plan.

**Comment by petrelharp** [comment]
> OK, so. I'm good with keepAllAncestralRetained=F if you guys like that. Shall I do a commit that adds this parameter to initializeTreeSeq(), and puts the value into a flag in SLiMSim for your use, and adds the doc above to the online doc in SLiMgui? I'm fine with just committing that and having it be a loose end for a couple days, until it gets hooked up to something real by a PR from you guys. Nobody will notice that it's even there, I imagine, and it won't break anything. I could do that now-ish if that sounds like a good plan.

I'll let @hyanwong answer this... but, probably? And, if you don't mind, could you do a PR instead of leaving something broken? We're working on pyslim, and the automatic tests for pyslim pull from the main branch of SLiM (as they need to, at least sometimes...).  Happy to advise on how to git this done.

**Comment by bhaller** [comment]
Sure, I'll do a PR if you like (but nothing would be broken, there would just be a new parameter for `initializeTreeSeq()` that doesn't actually do anything yet); but if so, then yes, please remind me how, sorry.  I'm spoiled by mostly working on my own in the SLiM codebase.  :-O  I'm on Slack now, if you want to walk me through it right now.  :->

**Comment by hyanwong** [comment]
> > OK, so. I'm good with keepAllAncestralRetained=F if you guys like that. Shall I do a commit that adds this parameter to initializeTreeSeq(), and puts the value into a flag in SLiMSim for your use, and adds the doc above to the online doc in SLiMgui? 
> I'll let @hyanwong answer this... but, probably? 

Sorry not to get round to this (it was late for me). I'm wondering if it is clearer to reverse the logic, so instead of `keepAllAncestralRetained=F` we have `retainCoalescentOnly=T` as the default. It seems a little less convoluted to me that way. Then we can say: "Individuals can be temporarily retained using `treeSeqRememberIndividuals(permanent=F)`, which means they are only retained if their genomes remain in the tree sequence. Genomes are always removed if not ancestral to the sample; if `retainCoalescentOnly=T` they are *also* removed if no longer a branch point (coalescent node, or MRCA) in the tree sequence. In some cases it may be useful to retain a node and its associated individual when it is simply an intermediate node in the ancestry (i.e. in the middle of a branch). This can be done by setting `retainCoalescentOnly=F`. In this case ancestral nodes that are intermediate ("unary nodes") and within a retained individual are kept, along with the retained individual itself."

**Comment by bhaller** [comment]
I do like `retainCoalescentOnly=T`; the way it has "retain" as the active verb, and describes what that verb implies, is more direct to me.  I guess the equivalent with the opposite sense would be `retainAllAncestral=F`.  But that latter flag is less descriptive; it immediately begs the question "as opposed to retaining... who, exactly?", whereas `retainCoalescentOnly=T` has a more obvious opposite, "as opposed to retaining everybody".  So I do lean toward `retainCoalescentOnly=T` for that reason.
However, I disagree that I'm the one with their finger on the pulse of the average SLiM user of tree-seq recording; I think that'd be @petrelharp.  :->  What do you think, Peter?

**Comment by petrelharp** [comment]
Well, if nothing well be broken, then go ahead with the code as you like, Ben.

And, the other suggestion for the flag name is fine, as long as it's clear right away from the docs that the ordinary use doesn't need to think about it. What I'd like to avoid is people without a strong mental model of simplification, etcetera thinking they need to start Retaining individuals to get what they need at the end, when all they need is the standard tree sequence. (This will happen!)

**Comment by bhaller** [comment]
OK, @hyanwong and @petrelharp, I have just committed the addition of `[l$ retainCoalescentOnly=T]` to `initializeTreeSeq()`.  There is now a corresponding flag in `SLiMSim`, `retain_coalescent_only_`, that can be used when simplify is called; I guess it will translate (inversely?) to the new tskit flag.

The doc I wrote can be revised, of course.  For initializeTreeSeq() I wrote:

> The retainCoalescentOnly parameter controls how, exactly, simplification of the tree-sequence data is performed in SLiM (both for auto-simplification and for calls to treeSeqSimplify()).  The default of T should work for almost all users of tree-sequence recording; changing it to F may dramatically increase memory usage and runtime, without benefit for most users.  More specifically, this parameter controls the behavior of simplification for individuals and genomes that have been “retained” by calling treeSeqRememberIndividuals() with the parameter permanent=F.  See the documentation of treeSeqRememberIndividuals() for further discussion (section 24.13.2).

For the treeSeqRememberIndividuals() doc I adapted Yan's proposal.  Here's the full doc for the function, since I edited other parts of it to fit these changes too:

> – (void)treeSeqRememberIndividuals(object<Individual> individuals, [logical$ permanent = T])
> Mark the individuals specified by individuals to be kept across tree sequence table simplification.  This method may only be called if tree sequence recording has been turned on with initializeTreeSeq().  All currently living individuals are always kept across simplification; this method does not need to be called, and indeed should not be called, for that purpose.  Instead, treeSeqRememberIndividuals() allows any individual, including dead individuals, to be kept in the final tree sequence.  Typically this would be used, for example, to keep particular individuals that you wanted to be able to trace ancestry back to in later analysis.  However, this is not the typical usage pattern for tree sequence recording; most models will not need to call this method.
> There are two ways to keep individuals across simplification.  If permanent is T (the default), then the specified individuals will be permanently remembered: their genomes will be added to the current sample, and they will always be present in the tree sequence.  Permanently remembering a large number of individuals will, of course, markedly increase memory usage and runtime.
> Supplying F for permanent will instead mark the individuals only for (temporary) retention: their genomes will not be added to the sample, and they will appear in the final tree sequence only if one of their genomes is retained across simplification.  In other words, the rule of thumb for retained individuals is simple: if a genome is kept by simplification, the genome’s corresponding individual is kept also, if it is retained.  Note that permanent remembering takes priority; calling this function with permanent=F on an individual that has previously been permanently remembered will not remove it from the sample.
> The behavior of simplification for individuals retained with permanent=F depends upon the value of the retainCoalescentOnly flag passed to initializeTreeSeq(); here we will discuss the behavior of that flag in detail.  First of all, genomes are always removed by simplification if they are not ancestral to the sample, unless their individual has been permanently remembered.  If retainCoalescentOnly is T (the default), they are also removed if they are not a branch point (i.e., a coalescent node or most recent common ancestor) in the tree sequence.  In some cases it may be useful to retain a genome and its associated individual when it is simply an intermediate node in the ancestry (i.e., in the middle of a branch).  This can be done by setting retainCoalescentOnly to F in your call to initializeTreeSeq().  In this case, ancestral genomes that are intermediate (“unary nodes”, in tskit parlance) and are within a retained individual are kept, along with the retained individual itself.  Since setting retainCoalescentOnly to F will prevent the unary nodes for retained individuals from being pruned, simplification may often be unable to prune very much at all from the tree sequence, and memory usage and runtime may increase rapidly.  This setting should therefore be used only with caution; it is rarely necessary.  See the pyslim documentation for further discussion of retaining and remembering individuals and the effects of the retainCoalescentOnly flag.
> The metadata (age, location, etc) that are stored in the resulting tree sequence are those values present at either (a) the final generation, if the individual is alive at the end of the simulation, or (b) the last time that the individual was remembered, if not.  Calling treeSeqRememberIndividuals() on an individual that is already remembered will cause the archived information about the remembered individual to be updated to reflect the individual’s current state; care should be taken to remember individuals at a point in time when their state is valid.  A case where this is particularly important is for the spatial location of individuals in continuous-space models.  SLiM automatically retains the portions of the genomes that comprise the first generation of any new subpopulation created with addSubpop() that are inherited by extant individuals, for easy recapitation and other analysis (see sections 17.2 and 17.10).  However, the individuals of the first generation are not remembered automatically, only their needed genomic information.

The tskit merge ought to build without code changes once Castedo's PR is accepted (which I see just happened a moment ago) and that version of tskit gets merged in, since it will add `vg->samples` back in.  I will go in and clean up the vargen usage as discussed later; that is orthogonal to this work, you can forget about that issue completely once you merge in the new tskit C API release that Ben is about to tag.

Progress!  :->

**Comment by hyanwong** [comment]
> OK, @hyanwong and @petrelharp, I have just committed the addition of `[l$ retainCoalescentOnly=T]` to `initializeTreeSeq()`

Thanks so much for this Ben.

> The doc I wrote can be revised, of course. For initializeTreeSeq() I wrote:
> 
> >The default of T should work for almost all users of tree-sequence recording; 

I'm not sure this is true, sadly. I think it *is* true if you are remembering (or retaining) everyone, or retaining every generation. If you are just retaining a few individuals, or only retaining every (say) 1000 generations, you should be fine. So how about:

> The retainCoalescentOnly parameter controls how, exactly, simplification of the tree-sequence data is performed in SLiM (both for auto-simplification and for calls to treeSeqSimplify()). More specifically, this parameter controls the behavior of simplification for individuals and genomes that have been “retained” by calling treeSeqRememberIndividuals() with the parameter permanent=F. The default of retainCoalescentOnly=T helps to keep the number of retained individuals relatively small, which is helpful if your simulation regularly flags many individuals for retaining. In this case, changing retainCoalescentOnly to F may dramatically increase memory usage and runtime, in a similar way to permanently Remembering all the individuals. See the documentation of treeSeqRememberIndividuals() for further discussion (section 24.13.2).

Also in the `treeSeqRememberIndividuals()` doc, how about:

> > ... Since setting retainCoalescentOnly to F will prevent the unary nodes for retained individuals from being pruned, simplification may often be unable to prune very much at all from the tree sequence, and memory usage and runtime may increase rapidly.  If you are retaining many individuals, this setting should therefore be used only with caution; it is not necessary if you are purely interested in the most recent common ancestors.  See the pyslim documentation for further discussion of retaining and remembering individuals and the effects of the retainCoalescentOnly flag.

I can change the docs, but I thought I'd ask here if these changes seem sensible. I do think that "rarely needed" is a bit of an overstatement: after all, for a while we did consider making `retainCoalescentOnly=F` the default, so it can't be *that* rare of a use case for individual retention.

**Comment by bhaller** [comment]
> > OK, @hyanwong and @petrelharp, I have just committed the addition of `[l$ retainCoalescentOnly=T]` to `initializeTreeSeq()`
> 
> Thanks so much for this Ben.
> 
> > The doc I wrote can be revised, of course. For initializeTreeSeq() I wrote:
> > > The default of T should work for almost all users of tree-sequence recording;
> 
> I'm not sure this is true, sadly. I think it _is_ true if you are remembering (or retaining) everyone, or retaining every generation. If you are just retaining a few individuals, or only retaining every (say) 1000 generations, you probably want `retainCoalescentOnly=F`. So how about: ...

Well, I think it is true that the default of T should work for almost all users of tree-sequence recording, because my guess is that almost all users of tree-sequence recording won't be retaining any individuals at all – so the value of this flag is irrelevant.  (That's why having the default be flipped would also be reasonable – for most users it actually doesn't matter at all, and it's just a question of what the better default behavior is for the small minority of users who do actually retain individuals.)  My goal was just to warn users as strongly as seemed reasonable, because of the concern @petrelharp expressed earlier.

That said, your edits seem clarifying and good, so the above point is moot; I just wanted to explain why I wrote it as I did.  I'll incorporate your changes now in a commit (I'll do it on my end given the way the docs get generated :->).  Thanks!

**Comment by bhaller** [comment]
(done)

**Comment by hyanwong** [comment]
Thanks. I'm working up a PR that actually implements this, with some tests.

**Comment by petrelharp** [comment]
Late arrival, but how about changing
> First of all, genomes are always removed by simplification if they are not ancestral to the sample, unless their individual has been permanently remembered.
to
```
First of all, genomes are always removed by simplification if they are not ancestral to the the final generation
and/or any permanently remembered genomes.
```

Otherwise, looks good!

**Comment by bhaller** [comment]
> Late arrival, but how about changing
> 
> > First of all, genomes are always removed by simplification if they are not ancestral to the sample, unless their individual has been permanently remembered.
> > to
> 
> ```
> First of all, genomes are always removed by simplification if they are not ancestral to the the final generation
> and/or any permanently remembered genomes.
> ```
> 
> Otherwise, looks good!

Only problem with this edit is: is an individual ancestral to itself?  If not (and my gut says no), this rewrite doesn't cover the case where the genome belongs to a remembered individual, but isn't *ancestral* to the final generation or any remembered individual.

So now I've got:

First of all, genomes are always removed by simplification if they are not ancestral to the final generation or any permanently remembered genomes, unless their individual has been permanently remembered.

But this is confusing and ugly.  I'm not seeing a simpler rewrite, though.  Any suggestions?

**Comment by bhaller** [comment]
How about:

First of all, genomes are always removed by simplification unless they are (a) part of the final generation, (b) ancestral to the final generation, (c) a genome of a permanently remembered individual, or (d) ancestral to a permanently remembered individual.

At least it's clear...

**Comment by bhaller** [comment]
OK, here's my proposed revision for this paragraph:

> The behavior of simplification for individuals retained with permanent=F depends upon the value of the retainCoalescentOnly flag passed to initializeTreeSeq(); here we will discuss the behavior of that flag in detail.  First of all, genomes are always removed by simplification unless they are (a) part of the final generation, (b) ancestral to the final generation, (c) a genome of a permanently remembered individual, or (d) ancestral to a permanently remembered individual.  If retainCoalescentOnly is T (the default), they are also always removed if they are not a branch point (i.e., a coalescent node or most recent common ancestor) in the tree sequence.  In some cases it may be useful to retain a genome and its associated individual when it is simply an intermediate node in the ancestry (i.e., in the middle of a branch).  This can be enabled by setting retainCoalescentOnly to F in your call to initializeTreeSeq().  In this case, ancestral genomes that are intermediate (“unary nodes”, in tskit parlance) and are within an individual that has been retained using the permanent=F flag here are kept, along with the retained individual itself.  Since setting retainCoalescentOnly to F will prevent the unary nodes for retained individuals from being pruned, simplification may often be unable to prune very much at all from the tree sequence, and memory usage and runtime may increase rapidly.  If you are retaining many individuals, this setting should therefore be used only with caution; it is not necessary if you are purely interested in the most recent common ancestors.  See the pyslim documentation for further discussion of retaining and remembering individuals and the effects of the retainCoalescentOnly flag.

I revised it a little in a few other spots, with the goal of clarifying that (a) retainCoalescentOnly=F does not, on its own, cause any additional genomes to be kept by simplification (you must also retain genomes), and (b) genomes (and individuals) get "retained" by passing permanent=F, a point that is made earlier too, but seems worth reinforcing here.  Hopefully this is good.  These concepts are very subtle and hard to explain accurately; diagrams will be worth their weight in gold.  I'll commit this doc change now.

**Comment by bhaller** [comment]
Hi @hyanwong.  I guess this is tied up together with the improvements to the testing framework?  And then there was a need for a small bug-fix in tskit that needed to get pulled into SLiM, too, right?  How is all that coming along?  Just checking in.  :->

**Comment by petrelharp** [comment]
Closed in #158!
